frontend/src/routes/rankings.jsx

```diff
@@ -0,0 +1,208 @@
+import { hideLoad, keepDate, showLoad } from "../features/part.js";
+import { generateIdentity, portraitProvider } from "../features/util.js";
+import Mistaken from "./mistaken.jsx";
+
+export default function Rankings() {
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
+  const location = useLocation();
+  const { y, m, d } = useParams();
+  const isweekly = location.pathname.endsWith("/week");
+  const pickdate = useSelector((state) => state.area.date);
+
+  const readDate = () => {
+    if (y && m && d) {
+      return `${y}-${m.toString().padStart(2, "0")}-${d.toString().padStart(2, "0")}`;
+    }
+    return pickdate || "";
+  };
+
+  const handleChange = (event) => {
+    const date = event.target.value;
+    dispatch(keepDate(date));
+
+    if (date) {
+      const [year, month, day] = date.split("-");
+      const weekSuffix = isweekly ? "/week" : "";
+      navigate(`/rankings/y/${year}/m/${parseInt(month)}/d/${parseInt(day)}${weekSuffix}`);
+    } else {
+      navigate("/rankings");
+    }
+  };
+
+  const params = {
+    ...(y && { y: parseInt(y) }),
+    ...(m && { m: parseInt(m) }),
+    ...(d && { d: parseInt(d) }),
+    ...(isweekly && { w: true }),
+  };
+
+  const { data: dict, isLoading, error } = useRetrieveRankingsQuery(params, { skip: false });
+
+  // Show or Hide LoadNote
+  useEffect(() => {
+    if (isLoading) {
+      dispatch(showLoad());
+    } else {
+      dispatch(hideLoad());
+    }
+  }, [isLoading, dispatch]);
+
+  if (error) {
+    return <Mistaken />;
+  }
+
+  if (isLoading || !dict) {
+    return null;
+  }
+
+  const showDate = () => {
+    if (Object.keys(params).length === 0) return "All time";
+
+    const date = new Date();
+    if (params.y) date.setFullYear(params.y);
+    if (params.m) date.setMonth(params.m - 1);
+    if (params.d) date.setDate(params.d);
+
+    const option = {};
+    if (params.y) option.year = "numeric";
+    if (params.m) option.month = "long";
+    if (params.d) option.day = "numeric";
+
+    return date.toLocaleDateString("en-US", option);
+  };
+
+  const showHead = () => {
+    let name = "";
+    if (params.w) name = "Weekly";
+    else if (params.d) name = "Daily";
+    else if (params.m) name = "Monthly";
+    else if (params.y) name = "Yearly";
+    else name = "All time";
+    return name;
+  };
+
+  const readLink = (conf = "") => {
+    const date = new Date();
+    let link = "";
+    if (conf === "d") link = `/rankings/y/${date.getFullYear()}/m/${date.getMonth() + 1}/d/${date.getDate()}`;
+    else if (conf === "w") link = `/rankings/y/${date.getFullYear()}/m/${date.getMonth() + 1}/d/${date.getDate()}/week`;
+    else if (conf === "m") link = `/rankings/y/${date.getFullYear()}/m/${date.getMonth() + 1}`;
+    else if (conf === "y") link = `/rankings/y/${date.getFullYear()}`;
+    else link = `/rankings`;
+    return link;
+  };
+
+  const scanDate = () => {
+    const today = new Date();
```

Comment:
It looks like this function is called many times in the rendering. Could it be just called once and the result stored in a variable?
Also, what happens if the page is left open for more than a day and the user goes back to it? The date will not be refreshed, right?